### DISCUSSION TITLE
主观测评不能正确运行

### DISCUSSION THREAD

**Discussion by MHHL-Frank** [open-compass/opencompass#1532] [Q&A] [answered]
使用opencompass进行主观测评，在容器里运行，无法连接huggingface，所以将模型下载后安装到容器，使用path指定路径，之前的客观测评运行都没有问题。
新配置了一个configs下的文件eval_qwen2-7B_subjective.py：
from mmengine.config import read_base
from opencompass.models import HuggingFacewithChatTemplate

with read_base():
    from .datasets.subjective.alignbench.alignbench_judgeby_critiquellm import alignbench_datasets

from opencompass.summarizers import AlignmentBenchSummarizer

datasets = alignbench_datasets
models = [
    dict(
        type=HuggingFacewithChatTemplate,
        abbr='qwen2-7b-instruct-hf',
        path='/home/jovyan/.cache/modelscope/hub/qwen/Qwen2-7B-Instruct',
        model_kwargs=dict(
            device_map='auto',
            trust_remote_code=True,
        ),
        tokenizer_kwargs=dict(
            padding_side='left',
            truncation_side='left',
            trust_remote_code=True,
        ),
        generation_kwargs=dict(
            do_sample=True,
        ),
        max_out_len=1024,
        batch_size=8,
        run_cfg=dict(num_gpus=1),
    )
]

在终端中运行python run.py configs/eval_qwen2-7B_subjective.py -r --debug

出现以下问题：
Traceback (most recent call last):
  File "/home/jovyan/opencompass/run.py", line 4, in <module>
    main()
  File "/home/jovyan/opencompass/opencompass/cli/main.py", line 351, in main
    runner(tasks)
  File "/home/jovyan/opencompass/opencompass/runners/base.py", line 38, in __call__
    status = self.launch(tasks)
  File "/home/jovyan/opencompass/opencompass/runners/local.py", line 131, in launch
    task.run()
  File "/home/jovyan/opencompass/opencompass/tasks/openicl_eval.py", line 114, in run
    self._score()
  File "/home/jovyan/opencompass/opencompass/tasks/openicl_eval.py", line 207, in _score
    icl_evaluator = ICL_EVALUATORS.build(self.eval_cfg['evaluator'])
  File "/home/jovyan/.conda/envs/opencompass/lib/python3.10/site-packages/mmengine/registry/registry.py", line 570, in build
    return self.build_func(cfg, *args, **kwargs, registry=self)
  File "/home/jovyan/.conda/envs/opencompass/lib/python3.10/site-packages/mmengine/registry/build_functions.py", line 121, in build_from_cfg
    obj = obj_cls(**args)  # type: ignore
TypeError: LMEvaluator.__init__() missing 2 required positional arguments: 'judge_cfg' and 'output_path'

目前的问题是不知道如何配置.py文件，上面的配置文件是根据各方的写法拼凑的。

**Comment by bittersweet1999** [comment]
主观评测需要用JudgeModel，你需要在config里写上一个Judge Model，参考 https://github.com/open-compass/opencompass/blob/main/configs/eval_subjective.py 这个config，指定好你要测的模型 https://github.com/open-compass/opencompass/blob/cfbd308edf2e2aa2b86e4275c7278dafaf611a44/configs/eval_subjective.py#L33，指定好要测的主观数据集https://github.com/open-compass/opencompass/blob/cfbd308edf2e2aa2b86e4275c7278dafaf611a44/configs/eval_subjective.py#L60，指定好你需要使用哪个Judge Model进行主观评测https://github.com/open-compass/opencompass/blob/cfbd308edf2e2aa2b86e4275c7278dafaf611a44/configs/eval_subjective.py#L70，为了确保代码一致，最好拉取一下最新的代码

**Reply by MHHL-Frank** [reply]
可以了，非常感谢！

**Comment by bittersweet1999** [comment, marked as answer]
更新了一下read me,供参考：
![图片](https://github.com/user-attachments/assets/8df6ee6b-7718-47b3-95d3-090113fc1e8b)

**Reply by MHHL-Frank** [reply]
我运行了一下，TypeError: AlignmentBenchSummarizer.__init__() got an unexpected keyword argument 'function'，将summarizer = dict(type=AlignmentBenchSummarizer, function='subjective')中的function='subjective'删掉后不报错，不知道是怎么回事？

**Reply by bittersweet1999** [reply]
不用改summarizer了，现在都直接使用from opencompass.summarizers import SubjectiveSummarizer，只要使用SubjectiveSummarizer这一个就可以了

**Reply by MHHL-Frank** [reply]
好的，谢谢！
